### ISSUE TITLE
Discussion: Adopting `float[]` over `ReadOnlyMemory<float>` for Vector Data Storage

### ISSUE THREAD

**Issue by r-Larch** [pgvector/pgvector-dotnet#33]
I've been exploring the `pgvector-dotnet` library and noticed that `Vector` uses `ReadOnlyMemory<float>` to store its data. I'm curious about the rationale behind this choice, especially considering the implications for performance and usability.

#### Drawbacks of Storing Data in `ReadOnlyMemory<float>`

One significant concern with `ReadOnlyMemory<float>` is its suitability for calculations. It appears that for nearly every operation that involves manipulation or computation, we're forced to convert the `ReadOnlyMemory<float>` into an array using `ToArray()`, which inherently causes a new array allocation. This behavior seems suboptimal from a performance standpoint, especially in scenarios where high efficiency and low latency are crucial.

#### Observations from the Repository Internals

Delving into the code, it's evident that instances of the `Vector` type are created with either `float[]` or `string`. In cases where a `string` is used, it is promptly converted into a `float[]` before being stored as `ReadOnlyMemory<float>`. This practice further raises questions about the necessity of `ReadOnlyMemory<float>` when the data ultimately originates from or is converted to `float[]`.

https://github.com/pgvector/pgvector-dotnet/blob/ada5859d4da1b60cad48a7c42396afa4a167870f/src/Pgvector/Vector.cs#L7-L16

#### Proposal for Discussion

Considering the aforementioned points, I propose a discussion on the potential advantages of directly using `float[]` for storing vector data. This change could simplify the codebase, reduce unnecessary allocations, and potentially enhance the library's performance. I'm interested in hearing the community's thoughts on this matter and understanding the original considerations that led to the choice of `ReadOnlyMemory<float>`.

**Comment by r-Larch** [comment]
For those looking to access `Vector` data as an array without copying, here's a workaround using an extension method that leverages `MemoryMarshal.TryGetArray` to retrieve the underlying `float[]` directly from a `Vector` instance. This approach avoids the performance cost of memory allocation associated with `ToArray()`.

```csharp
using System.Runtime.InteropServices;
using Vector = Pgvector.Vector;

public static class VectorExtensions {
    public static float[] AsArray(this Vector vector)
    {
        if (MemoryMarshal.TryGetArray(vector.Memory, out var seg)) {
            return seg.Array!;
        }
        throw new Exception("Failed to get Array");
    }
}
```

<details>
<summary>Test Code</summary>

To ensure this method works as expected, here's a test demonstrating it retrieves the original array reference without copying:

```csharp
[Fact]
public void Test()
{
    var arr = new float[] { 1, 2, 3, 4 };
    var vector = new Vector(arr);

    var arr2 = vector.AsArray();

    Assert.Equal(arr.Length, arr2.Length);
    Assert.Equal(arr, arr2);
    Assert.True(arr.SequenceEqual(arr2));
    Assert.True(object.ReferenceEquals(arr, arr2));
}
```

</details>

**Comment by roji** [comment]
> One significant concern with ReadOnlyMemory<float> is its suitability for calculations. It appears that for nearly every operation that involves manipulation or computation, we're forced to convert the ReadOnlyMemory<float> into an array using ToArray(), which inherently causes a new array allocation.

Which operations are you referring to exactly? Many/most operations involving float-related calculations already accept `ReadOnlySpan<float>` (and therefore already accept a `ROM<float>` directly); for instance, that's the case with the TensorPrimitives operations being introduced into .NET itself.

In fact, the main thinking behind accepting ROM<float>` is precisely to **avoid** unneeded copying of memory. For example, ROM can be used as a slice over a larger chunk of memory, where e.g. multiple segments are laid out sequentially; if an API accepts an array rather than a ROM, you must copy that segment out to an array. Also, ROM allows passing in unmanaged memory (possibly returned by a call into a native library), or stack-allocated memory (for cases where the data is relatively small and bounded, and it's desirable to avoid heap allocation because of the added GC pressure).

In short, ROM provides more flexibility for the types of inputs it allows (slices, unmanaged or stack memory), thereby reducing the need for copying. I'd be very interested in concrete examples of where this causes **additional** allocations.

> In cases where a string is used, it is promptly converted into a float[] before being stored as ReadOnlyMemory<float>.

That's normal and perfectly fine - in this very particular case (conversion from string), we indeed use an array to actual float data after converting it. That is then cast into a ROM (costless operation - there's no copying going on).

/cc @stephentoub
